### PR TITLE
Update ai sample code

### PR DESCRIPTION
**Pull request by DevAakif**
- Upgraded OpenAI models to latest versions
- Migrated from text completions to chat completions
- Updated Twitter connector configuration
- Updated Gmail connector functions
- Replaced deprecated Stable Diffusion models

All these changes are tested

Resolves https://github.com/ballerina-platform/ballerina-dev-website/issues/8648

### CODE REVIEW

File: audio_to_text_summarization_using_openai/main.bal [hunk 1]
```diff
@@ -53,17 +47,17 @@ public function main(string podcastURL) returns error? {
...
     if summerizedText is () {
 	    return error("Failed to summarize the given audio.");
     }
     io:println("Summarized text: ", summerizedText);
  
     // Tweet it out!
-    twitter:Client twitter = check new (twitterConfig);
-    var tweet = check twitter->tweet(summerizedText);
+    final twitter:Client twitter = check new({auth: {token:token}});
+    twitter:TweetCreateResponse tweet = check twitter->/tweets.post(payload = {text: summerizedText});
```
Comment (NipunaRanasinghe) [inline audio_to_text_summarization_using_openai/main.bal:61]: do we need named args in here? Can't we do
```suggestion
    twitter:TweetCreateResponse tweet = check twitter->/tweets.post({text: summerizedText});
```
?

File: compare_similarity_between_two_texts_using_openai/main.bal [hunk 2]
```diff
@@ -5,15 +5,15 @@ import ballerinax/openai.embeddings;
 configurable string openAIToken = ?;
 
 public function main() returns error? {
-    final embeddings:Client openAIEmbeddings = check new ({auth: {token: openAIToken}});
+    final embeddings:Client embeddingsClient = check new ({auth: {token: openAIToken}});
```
Comment (NipunaRanasinghe) [inline compare_similarity_between_two_texts_using_openai/main.bal:8]: Let's not change the client naming conventions we've used through out the samples
```suggestion
    final embeddings:Client openAIEmbeddings = check new ({auth: {token: openAIToken}});
```

File: audio_to_text_summarization_using_openai/main.bal [hunk 3]
```diff
@@ -1,38 +1,28 @@
...
-    apiSecret: twitterApiSecret,
-    accessToken: twitterAccessToken,
-    accessTokenSecret: twitterAccessTokenSecret
-};
 
 public function main(string podcastURL) returns error? {
     // Creates a HTTP client to download the audio file
     http:Client podcastEP = check new (podcastURL);
-    http:Response httpResp = check podcastEP->/get();
+    http:Response httpResp = check podcastEP->get("");
```
Comment (NipunaRanasinghe) [inline audio_to_text_summarization_using_openai/main.bal:19]: whats the reason for this change?

File: correct_grammar_and_spelling_in_text_using_openai/main.bal [hunk 4]
```diff
@@ -10,15 +10,18 @@ public function main(string filePath) returns error? {
...
 
-    text:CreateEditRequest editReq = {
-        input: check io:fileReadString(filePath),
-        instruction: "Fix grammar and spelling mistakes.",
-        model: "text-davinci-edit-001"
+    chat:CreateChatCompletionRequest request = {
+        model: "gpt-4o-mini",
+        messages: [{
+            "role": "user",
+            "content": string `Fix grammar and spelling mistakes of the content ${check io:fileReadString(filePath)}`
```
Comment (NipunaRanasinghe) [inline correct_grammar_and_spelling_in_text_using_openai/main.bal:19]:
```suggestion
            "content": string `Fix grammar and spelling mistakes of the content: ${check io:fileReadString(filePath)}`
```

File: create_and_send_customized_greeting_cards_using_openai/service.bal [hunk 5]
```diff
@@ -18,21 +19,25 @@ type GreetingDetails record {|
...
-                    prompt: string `Generate a greeting for a/an ${occasion}.${"\n"}Special notes: ${specialNotes}`,
-                    model: "text-davinci-003",
-                    max_tokens: 100
+                worker greetingWorker returns string|error? {
+                chat:CreateChatCompletionRequest request = {
+                    model: "gpt-4o-mini",
+                    messages: [{
+                        "role": "user",
+                        "content": string `Generate a greeting for a/an ${occasion}.${"\n"}Special notes: ${specialNotes}`
+                    }]
```
Comment (NipunaRanasinghe) [inline create_and_send_customized_greeting_cards_using_openai/service.bal:34]: incorrect formatting

**Review by NipunaRanasinghe — CHANGES_REQUESTED** [review]

File: audio_to_text_summarization_using_openai/main.bal (hunk 1 above)
Comment (DevAakif) [reply audio_to_text_summarization_using_openai/main.bal:61]: > do we need named args in here? Can't we do
> 
> ?

We can. I'll make the necessary change

File: discord_thread_summarizer_using_openai/ai_utils.bal [hunk 6]
```diff
@@ -8,11 +8,11 @@ final chat:Client chatClient = check new ({
...
-function generateChatCompletion(string prompt, string model = "gpt-3.5-turbo") returns string|error {
+function generateChatCompletion(string prompt, string model = "gpt-4o-mini") returns string|error {
     chat:CreateChatCompletionResponse res = check chatClient->/chat/completions.post({
         model,
         messages: [{role: "user", content: prompt}]
     });
     string? content = res.choices[0]?.message?.content;
     return content ?: error("The message is empty.");
-}
+}
```
Comment (xlight05) [inline discord_thread_summarizer_using_openai/ai_utils.bal:18]: Add newline

**Comment by xlight05** [comment]
OpenAI models related changes looks good for me =)

File: audio_to_text_summarization_using_openai/main.bal (hunk 3 above)
Comment (DevAakif) [reply audio_to_text_summarization_using_openai/main.bal:19]: > whats the reason for this change?

The reason for this is that it is the only way it is currently supported.
For your reference: https://discord.com/channels/957996897782616114/1286605582022873141/1286607360336334850

File: convert_audio_to_text_and_translate_using_openai/main.bal [hunk 7]
```diff
@@ -26,10 +26,22 @@ public function main(string audioURL, string toLanguage) returns error? {
     audio:CreateTranscriptionResponse transcriptionRes = check openAIAudio->/audio/translations.post(translationsReq);
     io:println("Audio text in English: ", transcriptionRes.text);
 
+
```
Comment (NipunaRanasinghe) [inline convert_audio_to_text_and_translate_using_openai/main.bal:29]:
```suggestion
```

File: create_and_send_customized_greeting_cards_using_openai/service.bal [hunk 8]
```diff
@@ -18,28 +19,32 @@ type GreetingDetails record {|
...
+                    return response.choices[0].message.content;
+                }
+
+                worker imageWorker returns string|error? {
+                    images:CreateImageRequest imagePrompt = {
+                        prompt: string `Greeting card design for ${occasion}, ${specialNotes}`
+                    };
+                    images:ImagesResponse imageRes = check openAIImages->/images/generations.post(imagePrompt);
+                    return imageRes.data[0].url;
+                }
```
Comment (NipunaRanasinghe) [inline create_and_send_customized_greeting_cards_using_openai/service.bal:47]: looks like the formatting is incorrect. @DevAakif please use Ballerina VSCode formatter

File: create_products_in_shopify_using_openai/main.bal [hunk 9]
```diff
@@ -23,14 +24,19 @@ service / on new http:Listener(9090) {
         sheets:Range range = check gsheets->getRange(googleSheetId, "Sheet1", "A2:F");
         var [name, benefits, features, productType] = getProduct(range);
 
+
```
Comment (NipunaRanasinghe) [inline create_products_in_shopify_using_openai/main.bal:27]: additional newline
```suggestion
```

File: generate_and_send_a_creative_email_using_openai_and_stability_ai/main.bal [hunk 10]
```diff
@@ -18,21 +18,27 @@ public type EmailDetails record {|
...
+                    "content": string `Generate a creative poem on the topic ${emailDetails.topic}.`
+                }],
                 max_tokens: 1000
             };
-            text:CreateCompletionResponse completionRes = check openAIText->/completions.post(textPrompt);
-            return completionRes.choices[0].text;
+
+            chat:CreateChatCompletionResponse response = check openAIChat->/chat/completions.post(request);
+            return response.choices[0].message.content;
+
```
Comment (NipunaRanasinghe) [inline generate_and_send_a_creative_email_using_openai_and_stability_ai/main.bal:35]: ditto
```suggestion
```

File: insert_data_to_vector_db_weaviate/main.bal [hunk 11]
```diff
@@ -48,7 +48,9 @@ public function main() returns error? {
     foreach int i in 0 ... embeddingResponse.data.length() - 1 {
         documentObjects[i].vector = embeddingResponse.data[i].embedding;
     }
+
     final weaviate:Client weaviate = check new ({auth: {token: weaviateToken}, timeout: 100}, weaviateURL);
+    io:println("weaviate client initialized: ", weaviate);
```
Comment (NipunaRanasinghe) [inline insert_data_to_vector_db_weaviate/main.bal:53]: no need to print the client object in here IMO
```suggestion
```

File: question_answering_based_on_context_using_openai/main.bal [hunk 12]
```diff
@@ -33,12 +33,18 @@ service / on new http:Listener(8080) {
...
-        text:CreateCompletionRequest prmt = {
-            prompt: prompt,
-            model: "text-davinci-003"
+
+            chat:CreateChatCompletionRequest request = {
+            model: "gpt-4o-mini",
+            messages: [{
+                "role": "user",
+                "content": prompt
+                }]
```
Comment (NipunaRanasinghe) [inline question_answering_based_on_context_using_openai/main.bal:42]: incorrect formatting

File: question_answering_based_on_context_using_openai/main.bal [hunk 13]
```diff
@@ -33,12 +33,18 @@ service / on new http:Listener(8080) {
...
         };
-        text:CreateCompletionResponse completionRes = check openAIText->/completions.post(prmt);
-        return completionRes.choices[0].text;
+
+        chat:CreateChatCompletionResponse response = check openAIChat->/chat/completions.post(request);
+        return response.choices[0].message.content;
+
```
Comment (NipunaRanasinghe) [inline question_answering_based_on_context_using_openai/main.bal:47]:
```suggestion
```

File: question_answering_based_on_context_using_openai_and_pinecone/main.bal [hunk 14]
```diff
@@ -3,8 +3,8 @@ import ballerina/io;
 import ballerina/regex;
 import ballerinax/googleapis.sheets;
 import ballerinax/openai.embeddings;
-import ballerinax/openai.text;
 import ballerinax/pinecone.vector as pinecone;
+import ballerinax/openai.chat;
```
Comment (NipunaRanasinghe) [inline question_answering_based_on_context_using_openai_and_pinecone/main.bal:7]: incorrect imports order. Please use Ballerina VSCode formatter on all source files

File: question_answering_based_on_context_using_openai_and_pinecone/main.bal [hunk 15]
```diff
@@ -42,15 +42,20 @@ service / on new http:Listener(8080) {
...
+                "role":"user",
+                "content":prompt
+            }]
         };
-        text:CreateCompletionResponse completionRes = check openAIText->/completions.post(prmt);
-        return completionRes.choices[0].text;
+
+        chat:CreateChatCompletionResponse response = check openAIChat->/chat/completions.post(request);
+        return response.choices[0].message.content;
+
```
Comment (NipunaRanasinghe) [inline question_answering_based_on_context_using_openai_and_pinecone/main.bal:56]:
```suggestion
```

File: slackbot_using_azure_openai/service.bal [hunk 16]
```diff
@@ -49,19 +49,24 @@ service /slack on new http:Listener(8080) {
...
+
+        chat:ChatCompletionResponseMessage? response = choices[0].message;
         string? responseText = response?.content;
         if response is () || responseText is () {
             return error("Error in response generation");
         }
 
-        history.push({role: ASSISTANT, content: response.content});
+
+        history.push({role: ASSISTANT, content: requestText});
```
Comment (NipunaRanasinghe) [inline slackbot_using_azure_openai/service.bal:69]:
```suggestion
        history.push({role: ASSISTANT, content: requestText});
```

File: tweet_on_upcoming_and_recently_released_movies_using_azure_openai/main.bal [hunk 17]
```diff
@@ -28,18 +28,18 @@ public function main() returns error? {
         prompt += string `${i}. ${upcomingMovies.results[i - 1].title} `;
     }
 
-    text:Deploymentid_completions_body completionsBody = {
-        prompt,
-        max_tokens: MAX_TOKENS
+    final twitter:Client twitter = check new({
+    auth: { token:token}});
```
Comment (NipunaRanasinghe) [inline tweet_on_upcoming_and_recently_released_movies_using_azure_openai/main.bal:32]: incorrect formatting

**Review by NipunaRanasinghe — CHANGES_REQUESTED** [review]

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

**Pull request merged**